Nyx cycle changes

mskinner5278 opened the pull request:
This pull request should be reviewed after PR 309.

JunAishima commented on daq_macros.py:
```diff
@@ -2259,11 +2283,11 @@ def gotoMaxRaster(rasterResult,multiColThreshold=-1,**kwargs):
-      beamline_lib.mvaDescriptor("sampleX",x,
-                                 "sampleY",y,
-                                 "sampleZ",z,
+      beamline_lib.mvaDescriptor("sampleX",x/1000,
```
there are a lot of these unit changes - can you abstract them so that they don't make destructive changes to AMX/FMX?

mskinner5278 replied:
These changes are being made directly because they will not be kept in LSDC after the shutdown. The MD2 code for rastering will be completely different, and none of these changes are expected to remain.

JunAishima commented on daq_macros.py:
```diff
@@ -3258,7 +3290,37 @@ def loop_center_xrec():
+    logger.info(f'taking image at {omega_target}')
+    timeout = 5
+    # change image mode to single(0)
+    beamline_support.setPvValFromDescriptor("lowMagImMode",0)
```
is there a reason why you this function name instead of setPvDesc? that alias is defined in the imports and makes it significantly easier to read

mskinner5278 replied:
I have fixed this with [d4fadc2](https://github.com/NSLS-II/lsdc/pull/314/commits/d4fadc2c96a80c281c5981a674a6373d791f4b6b)

JunAishima commented on daq_main_common.py:
```diff
@@ -40,8 +40,8 @@ def pybass_init():
   daq_lib.init_var_channels()
-  #if getBlConfig(config_params.DETECTOR_OBJECT_TYPE) == config_params.DETECTOR_OBJECT_TYPE_LSDC:
-  det_lib.init_detector()  
+  if getBlConfig(config_params.DETECTOR_OBJECT_TYPE) != config_params.DETECTOR_OBJECT_TYPE_NO_INIT:
```
not sure why NO_INIT is defined - can't you make use of one of the other states?

JunAishima commented on gui/control_main.py:
```diff
@@ -353,6 +353,8 @@ def createSampleTab(self):
         self.closeShutterButton = QtWidgets.QPushButton("Close Photon Shutter")
         self.closeShutterButton.clicked.connect(self.closePhotonShutterCB)
         self.closeShutterButton.setStyleSheet("background-color: red")
+        self.parkRobotButton = QtWidgets.QPushButton("Park Robot")
```
you need to find a way to make these activations and deactivations clean for NYX (or whatever other state you want to call it). these must not appear in code that is to run on AMX/FMX.
this problem happens throughout the GUI code - both the deactivations and new buttons

JunAishima commented on gui/control_main.py:
```diff
@@ -3265,22 +3347,22 @@ def getCurrentFOV(self):
     def screenXPixels2microns(self, pixels):
         fov = self.getCurrentFOV()
-        fovX = fov["x"]
+        fovX = fov["x"]*1000
```
again, these factors of 1000 need to be extracted out somewhere.

mskinner5278 replied:
This code will not remain for the MD2.

JunAishima commented on start_bs.py:
```diff
@@ -105,9 +105,9 @@ def filter_camera_data(camera):
-    x = Cpt(EpicsMotor, ':GX}Mtr')
-    y = Cpt(EpicsMotor, ':PY}Mtr')
-    z = Cpt(EpicsMotor, ':PZ}Mtr')
+    x = Cpt(EpicsMotor, ':X}Mtr')
```
this seems like a good place to extract a configuration variable - it cannot remain here

mskinner5278 replied:
This code will be different for the MD2. I originally tried to make this part configurable, but did not have the time.

JunAishima commented on daq_macros.py:
```diff
@@ -815,7 +819,7 @@ def runDozorThread(directory,
     global rasterRowResultsList,processedRasterRowCount
 
-    time.sleep(0.5) #allow for file writing
+    time.sleep(1.0) #allow for file writing
```
this is another area where the different behavior should not be propagated to other beamlines

mskinner5278 replied:
It will be re-tested when the MD2 rastering is being tested.

JunAishima reviewed (changes requested):
there are generally a lot of hard-coded changes that cannot be brought into LSDC because they would change behavior and look for AMX/FMX. please change these areas to enable merging into the main LSDC codebase

mskinner5278 commented:
To summarize, any changes that look like they would directly interfere with AMX/FMX rastering will not remain because it is not going to be used for the MD2 rastering code. We will likely split the rastering code into a different function to make the call with exporter. Those changes were never ever intended to be brought into master, and were not written to be.

JunAishima commented:
we have agreed that to expedite the code review process, NYX-specific changes that are will affect AMX/FMX LSDC will be handled following the full approve of NYX for proprietary users.

tacaswell commented on daq_macros.py:
```diff
@@ -1867,17 +1871,23 @@ def snakeRasterBluesky(rasterReqID, grain=""):
-        raster_flyer.zebra.reset.put(1)  # reset after every row to make sure it is clear for the next row
-        time.sleep(0.2)  # necessary for reliable row processing - see comment in commit 6793f4
+        #raster_flyer.zebra.reset.put(1)  # reset after every row to make sure it is clear for the next row
+        time.sleep(0.3)  # necessary for reliable row processing - see comment in commit 6793f4
```
use `yield from bps.sleep(...)` in plans.  Putting time.sleep will block the event loop in bad ways.

mskinner5278 replied:
fixed with [d706a4b](https://github.com/NSLS-II/lsdc/pull/314/commits/d706a4b190cc4c4a7169dfeaaa553272534189a8)

tacaswell commented on daq_macros.py:
```diff
@@ -1867,17 +1871,23 @@ def snakeRasterBluesky(rasterReqID, grain=""):
+        logger.info(f'starting new row: {row_index}')
+        zMotAbsoluteMove, zEnd, yMotAbsoluteMove, yEnd, xMotAbsoluteMove, xEnd = raster_positions(row, stepsize, omegaRad+90, rasterStartZ*1000, rasterStartY*1000, rasterStartX*1000, row_index)
+        vector = {'x': (xMotAbsoluteMove/1000, xEnd/1000), 'y': (yMotAbsoluteMove/1000, yEnd/1000), 'z': (zMotAbsoluteMove/1000, zEnd/1000)}
+        yield from bps.mv(samplexyz.x, xMotAbsoluteMove/1000)
```
If it is safe to move in parallel, these can all be put in a single `bps.mv(mtr, target, mtr, target, ...)` call which should save some time.

mskinner5278 replied:
Great idea, I have put that in with [59cb602](https://github.com/NSLS-II/lsdc/pull/314/commits/59cb602fd08c29ca22d01515ba775b17798eadb2)

tacaswell commented on daq_macros.py:
```diff
@@ -1867,17 +1871,23 @@ def snakeRasterBluesky(rasterReqID, grain=""):
-        xMotAbsoluteMove, xEnd, yMotAbsoluteMove, yEnd, zMotAbsoluteMove, zEnd = raster_positions(row, stepsize, omegaRad, rasterStartX, rasterStartY, rasterStartZ, row_index)
-        vector = {'x': (xMotAbsoluteMove, xEnd), 'y': (yMotAbsoluteMove, yEnd), 'z': (zMotAbsoluteMove, zEnd)}
+        logger.info(f'starting new row: {row_index}')
+        zMotAbsoluteMove, zEnd, yMotAbsoluteMove, yEnd, xMotAbsoluteMove, xEnd = raster_positions(row, stepsize, omegaRad+90, rasterStartZ*1000, rasterStartY*1000, rasterStartX*1000, row_index)
```
What are the `*1000` and `/1000` for ?!

mskinner5278 replied:
NYX has some units with different scaling, so we need to convert between microns and millimeters in several places to use the AMX/FMX code. I think there's still a map of it on the wall of my cubicle.

mskinner5278 commented on daq_macros.py:
```diff
@@ -1867,17 +1871,23 @@ def snakeRasterBluesky(rasterReqID, grain=""):
       return
-    raster_flyer.configure_detector(file_prefix=rasterFilePrefix, data_directory_name=data_directory_name)
-    raster_flyer.detector.stage()
+    #raster_flyer.configure_detector(file_prefix=rasterFilePrefix, data_directory_name=data_directory_name)
```
I originally left them in so the code could eventually be made safe to merge into master (probably creating another if beamline==nyx check), but changing to the MD2 means that will never need to happen.

mskinner5278 replied:
But, since we do not plan to do that, I have now removed them. [24130d1](https://github.com/NSLS-II/lsdc/pull/314/commits/24130d1029ce6c5abd5f04bd3219700a5e9e08c6)

JunAishima commented:
> None of this looks like it has any impact on data security?

Please see my comment just below yours - there seems to be only one place where data security could be an issue within this pull request. The comment refers to a new function used to write out images and is implicitly using the current working directory instead of explicitly using the visitDirectory

mskinner5278 commented on daq_macros.py:
```diff
@@ -3260,7 +3290,37 @@ def loop_center_xrec():
+    # start camera
+    beamline_support.setPvValFromDescriptor("lowMagAcquire",1)
+    try:
+      with urllib.request.urlopen(daq_utils.lowMagCamURL, timeout=timeout) as response:
```
You are correct, I have now added the visit directory call with [59cb602](https://github.com/NSLS-II/lsdc/pull/314/commits/59cb602fd08c29ca22d01515ba775b17798eadb2)

JunAishima commented:
@mskinner5278 do you think you have addressed all of the issues with the pull request that have been brought up?

mskinner5278 commented:
> @mskinner5278 do you think you have addressed all of the issues with the pull request that have been brought up?

Yep, let me know if you want more changes added.

tacaswell commented on daq_macros.py:
```diff
@@ -3260,7 +3285,37 @@ def loop_center_xrec():
+        logger.info("xnz: read")
+        image_data = response.read()
+        logger.info("xnz: read")
+      with open(getBlConfig("visitDirectory")+"findloop_"+str(omega_target/40)+".jpg", "wb") as filename:
```
```suggestion
      with open(getBlConfig("visitDirectory")+"findloop_"+str(omega_target//40)+".jpg", "wb") as filename:
```

Be sure you get an int.

tacaswell reviewed (approved):
I'm approving this with reservations.

JunAishima reviewed (approved):
I am content that any data security issues have been resolved.
